`opam switch create` should check whether cached sources are up to date

```
opam switch create catenable ocaml-base-compiler.4.07.1 
```

failed with output (trimmed) like this:

```
<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ocaml-base-compiler.4.07.1] found in cache
...
[ERROR] The compilation of ocaml-base-compiler failed at
        "/home/???/.opam/opam-init/hooks/sandbox.sh build make world".
...
# /usr/bin/ld: backtrace.pic.o:/home/???/.opam/catenable/.opam-switch/build/ocaml-base-compile
r.4.07.1/byterun/backtrace.c:31: multiple definition of `caml_debug_info'; backtrace_prim.pic.o:/hom
e/???/.opam/catenable/.opam-switch/build/ocaml-base-compiler.4.07.1/byterun/backtrace_prim.c:4
7: first defined here                                                                              
# collect2: error: ld returned 1 exit status
# make[1]: *** [Makefile:208: libcamlrun_shared.so] Error 1
# make[1]: Leaving directory '/home/???/.opam/catenable/.opam-switch/build/ocaml-base-compiler
.4.07.1/byterun'                                                                                   
# make: *** [Makefile:389: coldstart] Error 2
```

I was confused because it looked exactly like [this bug][], which had been fixed months ago.

[this bug]: https://github.com/ocaml/opam-repository/pull/16722

The problem went away after I did this:

```
opam update -a
opam upgrade
```

I'm guessing that `opam switch create` used the sources it found in my cache without checking whether they were up to date with the repo.  It only picked up the patch after I instructed the `update` and `upgrade`.  Does that sound plausible?

If so, would it be possible to make opam check whether its cached sources are up to date before using them?

```
# opam config report
# opam-version      2.0.0 
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=debian os-version=10.0
[ERROR] No switch is currently set. Please use 'opam switch' to set or install a switch
# read-state        OpamStd.OpamSys.Exit(50)
```

As an aside (I know this is not your fault), I encountered this error while following this [quick start guide][] for my first dune project. It was pretty discouraging to start my first dune project with two hours investigating a link error in the compiler, followed by an hour [setting up automatic formatting][], followed by an hour investigating and then working around a type error in [code generated by a ppx][].

[quick start guide]: https://ocamlverse.github.io/content/quickstart_ocaml_project_dune.html
[setting up automatic formatting]: https://github.com/ocaml/dune/issues/3854
[code generated by a ppx]: https://github.com/janestreet/base_quickcheck/issues/2 

If you're wondering why I started my first project with old compiler sources in my opam cache, I think they were left over from the last time I tried to set up OCaml, which resulted in my filing issues with the [ocaml][] and [opam][] repos.  :-(

[ocaml]: https://github.com/ocaml/ocaml.org/issues/979
[opam]: https://github.com/ocaml/opam/issues/3282

Luckily, I have had some positive experiences with OCaml before, so I am not so easily discouraged. :-)

--- a/master_changes.md
+++ b/master_changes.md
@@ -27,6 +27,7 @@ New option/command/subcommand are prefixed with ◈.
 ## Switch
   * Fix `--update-invariant` when removing or changing package name [#4360 @AltGr - fix #4353]
   * Fix updates of the invariant with `--update-invariant` [#4431 @AltGr]
+  * Add a message to advise update if not done since 3 weeks [#4415 @rjbou - fix #4377]
 
 ## Pin
   * Url pin: fix opamfile format upgrade [#4366 @rjbou - fix #4365]

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -133,17 +133,7 @@ let get_installed_atoms t atoms =
 (* Check atoms for pinned packages, and update them. Returns the state that
    may have been reloaded if there were changes *)
 let update_dev_packages_t ?(only_installed=false) atoms t =
-  (* Check last update of the repo *)
-  let last_update =
-    (Unix.stat (OpamFilename.to_string
-                  (OpamPath.state_cache
-                     (OpamStateConfig.(!r.root_dir))))).Unix.st_mtime
-  in
-  let too_old = float_of_int (3600*24*21) in
-  if (Unix.time () -. last_update) > too_old then
-    OpamConsole.note "It seems you have not updated your repositories \
-                      for a while. Consider updating them with:\n%s\n"
-      (OpamConsole.colorise `bold "opam update");
+  OpamRepositoryState.check_last_update ();
   if OpamClientConfig.(!r.skip_dev_update) then t else
   let working_dir = OpamClientConfig.(!r.working_dir || !r.inplace_build) in
   let to_update =

--- a/src/client/opamSwitchCommand.ml
+++ b/src/client/opamSwitchCommand.ml
@@ -204,6 +204,7 @@ let install_compiler ?(additional_installs=[]) ?(deps_only=false) t =
      OpamEnv.check_and_print_env_warning t);
     t
   end else
+  let () = OpamRepositoryState.check_last_update () in
   let atoms = OpamFormula.atoms invariant in
   let names_of_atoms at = OpamPackage.Name.Set.of_list (List.map fst at) in
   let comp_roots = names_of_atoms atoms in

--- a/src/core/opamFilename.ml
+++ b/src/core/opamFilename.ml
@@ -168,6 +168,12 @@ let touch t =
 let chmod t p =
   Unix.chmod (to_string t) p
 
+let written_since file =
+  let last_update =
+    (Unix.stat (to_string file)).Unix.st_mtime
+  in
+  (Unix.time () -. last_update)
+
 let of_string s =
   let dirname = Filename.dirname s in
   let basename = Filename.basename s in

--- a/src/core/opamFilename.mli
+++ b/src/core/opamFilename.mli
@@ -254,6 +254,9 @@ val touch: t -> unit
 (** Change file permissions *)
 val chmod: t -> int -> unit
 
+(** Returns delay since last file update, based on mtime *)
+val written_since: t -> float
+
 (** Returns the closest parent of a directory (including itself) for which the
     predicate holds, if any *)
 val find_in_parents: (Dir.t -> bool) ->  Dir.t -> Dir.t option

--- a/src/state/opamRepositoryState.ml
+++ b/src/state/opamRepositoryState.ml
@@ -278,3 +278,14 @@ let write_config rt =
          if r.repo_url = OpamUrl.empty then None
          else Some (r.repo_url, r.repo_trust))
         rt.repositories)
+
+let check_last_update () =
+  let last_update =
+    OpamFilename.written_since
+      (OpamPath.state_cache (OpamStateConfig.(!r.root_dir)))
+  in
+  if last_update > float_of_int (3600*24*21) then
+    OpamConsole.note "It seems you have not updated your repositories \
+                      for a while. Consider updating them with:\n%s\n"
+      (OpamConsole.colorise `bold "opam update");
+

--- a/src/state/opamRepositoryState.mli
+++ b/src/state/opamRepositoryState.mli
@@ -103,3 +103,6 @@ val with_write_lock:
 
 (** Writes the repositories config file back to disk *)
 val write_config: rw repos_state -> unit
+
+(** Display a warning if repository has not been updated since 3 weeks *)
+val check_last_update: unit -> unit

